Prefer currently booted iOS Simulator

gregdenness opened the pull request:
This PR updates XCodeHelper.getSimulatorID to first check if a simulator is already running and use that device's UDID, if there are no simulators running it will fall back to original behaviour.

This improves usability when switching simulators manually via the Simulator app, as users may wish to switch simulators multiple times during testing. For example if I open an "iPad 11 inch" simulator specifically and then build I'd expect it to be installed and run on that simulator.

This change is non-breaking and falls back to the existing logic if no simulator is currently running.

gregdenness committed:
Prefer currently booted iOS Simulator

Updated XCodeHelper.getSimulatorID to check for a currently booted simulator via `xcrun simctl list devices booted` before falling back to project target flags or default logic. This improves behavior when switching simulators manually in the Simulator app.

joshtynjala commented on src/lime/tools/XCodeHelper.hx:
```diff
@@ -54,6 +54,21 @@ class XCodeHelper
+			// Match a line like: "    iPhone 15 Pro (E5B048F6-BE70-498E-B5F6-B84B8594DDBF) (Booted)"
+			var match = ~/.*\(([A-F0-9\-]{36})\)\s+\(Booted\)/;
+			if (match.match(line)) {
+				return match.matched(1);
```
For consistency with the rest of the code in this class, can you tweak this to use the existing `getSimulatorID()` method that we're using elsewhere? You can probably simplify the initial match as `line.indexOf("(Booted)")` in that case.

joshtynjala commented on src/lime/tools/XCodeHelper.hx:
```diff
@@ -173,6 +188,15 @@ class XCodeHelper
 	{
+		// try and get the simulator which is currently open and running
+		// if there are none running then fall back to getting the selected simulator
+		var booted = getBootedSimulatorID();
```
Can you also modify `getSimulatorName()` to also return the booted simulator name? Both `getSimulatorID()` and `getSimulatorName()` are expected to return the same simulator.

gregdenness replied:
Hi, yeah sure no problem. I'll do this next week. Thanks for the quick reply.

joshtynjala replied:
Great, thanks! It's a good enhancement.

gregdenness committed:
cleaned up code to be more consistent

getBootedSimulator now uses the same logic to match simulator name as elsewhere and also returns SimulatorInfo.
getSimulatorName now returns the booted simulator name.

gregdenness committed:
fixed a bug with extractSimulatorID

It appears that modern simulators have started adding their CPU types to their names, "like iPad Air (M4)" which was causing the exiting code to return the incorrect string for the ID (it would return 'M4' as the ID in that case). I fixed this by searching for ID's which are specifically 36 chars long which is the format the ID's follow.

I think this probably would have affected the code in the current release version of lime?

gregdenness commented:
I have updated the code to be consistent with the existing code as requested. I then encountered a strange issue where it wasn't working on certain simulators - those which had the cpu type in their names - such as "iPad air (M2)" - the (M2) part was breaking the extractSimulatorID function because of the parenthesis. I changed extractSimulatorID so that it looks for a 36 digit ID (all simulator IDs are). This is probably affecting the current lime release, though obviously only simulators which have the CPU type in their names.

joshtynjala commented:
Thanks! I'll test a bit and get it merged when I have a chance.

player-03 commented on src/lime/tools/XCodeHelper.hx:
```diff
@@ -43,15 +43,35 @@ class XCodeHelper
+	
+			var potentialID = line.substring(openParen + 1, closeParen);
+	
+			// Simulator IDs are always UUID strings exactly 36 chars long
```
If it really is that simple, we can save a lot of space here.

```haxe
private static function extractSimulatorID(line:String):String
{
	// Simulator IDs are always UUID strings exactly 36 chars long, in parentheses
	var id:EReg = ~/\(([A-F0-9\-]{36})\)/;
	return id.match(line) ? id.matched(1) : "";
}
```

gregdenness replied:
Yeah that's much shorter for sure.

player-03 commented on src/lime/tools/XCodeHelper.hx:
```diff
@@ -173,22 +207,32 @@ class XCodeHelper
+			return simulator;
 		}
-		return simulator.name;
+		return null;
```
`if (simulator != null) return simulator else return null` can be simplified to `return simulator`.
